projects/RabbitMQ.Client/client/impl/SocketFrameHandler.cs

```diff
@@ -274,22 +242,24 @@ public bool TryReadFrame(out InboundFrame frame)
 
         public async ValueTask SendHeaderAsync()
         {
-            _pipeWriter.Write(ProtocolHeader);
-            await _pipeWriter.FlushAsync().ConfigureAwait(false);
+            await _pipeWriter.WriteAsync(ProtocolHeader)
```

Comment:
I'd stick with the synchronous version to avoid the async overhead

Reply:
Is that overhead measurable? Those `PipeWriter` methods return a `ValueTask<T>`.

Reply:
> See [my comment](https://github.com/rabbitmq/rabbitmq-dotnet-client/pull/1347#discussion_r1395769821). If we'd like to further discuss, add a comment to the PR itself. We could always add a benchmark.

This is the way!

If the state machine becomes an issue, it's always possible, when only the last invocation is async, to just return the task. Although, that comes with its own [cost](https://blog.stephencleary.com/2016/12/eliding-async-await.html).

Reply:
The "only" difference is that `WriteAsync` calls `Write` and then returns `FlushAsync` :)

So by calling the Async version you are flushing after every write (which might be buffered and not actually be async), but ideally if we want to reap the benefits of ValueTask AND avoid the state machine you can assign the WriteAsync result to a variable, check if it has completed successfully, and only await it hasn't (i.e. is actually async).

WriteAsync:
```csharp
/// <summary>Writes the specified byte memory range to the pipe and makes data accessible to the <see cref="System.IO.Pipelines.PipeReader" />.</summary>
/// <param name="source">The read-only byte memory region to write.</param>
/// <param name="cancellationToken">The token to monitor for cancellation requests. The default value is <see cref="System.Threading.CancellationToken.None" />.</param>
/// <returns>A task that represents the asynchronous write operation, and wraps the flush asynchronous operation.</returns>
public virtual ValueTask<FlushResult> WriteAsync(ReadOnlyMemory<byte> source, CancellationToken cancellationToken = default)
{
    this.Write(source.Span);
    return FlushAsync(cancellationToken);
}
```

Reply:
@stebet thanks!

> deally if we want to reap the benefits of ValueTask AND avoid the state machine you can assign the WriteAsync result to a variable, check if it has completed successfully, and only await it hasn't (i.e. is actually async).

Ah I see now.

Reply:
Well, this is interesting. In the `SocketFrameHandler.WriteLoop` method I modified the code to always check `IsCompletedSuccessfully` and then `await` if not, and it caused `MassPublish` to lock up. I will add a TODO to the 7.0 release to look into "bypassing" `await` statements but for now I'm going to keep the code simple as it is.

Reply:
https://github.com/rabbitmq/rabbitmq-dotnet-client/issues/1413